### PR TITLE
[DPE-3590] Rel Changed Event only runs in its own peer relation hook

### PR DESCRIPTION
**Pull request by phvalguima**
The rolling-ops **MUST** guarantee that its peer relation event is executed within the peer relation hook context, to avoid any data drift between hooks. This PR enforces this behavior. If rolling-ops internal events are called in a different hook context, then we should process it differently.

In essence, any **READ** from the databag must be guaranteed to happen within the peer relation hook. **WRITE** can happen on other hooks. That means changes for all three events that belong to this lib.

This PR also removes the `process-locks` event, as they must be tightly coupled with its peer relation changed. This way, we guarantee the lock is only granted based on the up-to-date information from the peer-relation.

Likewise, we need a safe-guard in the `run-with-lock` event.

One comment in special, I believe we do not need the defer, as Marc pointed out: https://github.com/canonical/charm-rolling-ops/pull/10#discussion_r1493086374.

Closes: #13

### CODE REVIEW

File: lib/charms/rolling_ops/v0/rollingops.py [hunk 1]
```diff
@@ -335,9 +339,9 @@ def _on_relation_changed(self: CharmBase, event: RelationChangedEvent):
             self.charm.on[self.name].run_with_lock.emit()
 
         if self.model.unit.is_leader():
-            self.charm.on[self.name].process_locks.emit()
+            self._on_process_locks()
 
-    def _on_process_locks(self: CharmBase, event: ProcessLocks):
+    def _on_process_locks(self: CharmBase):
```
Comment (marcoppenheimer) [inline lib/charms/rolling_ops/v0/rollingops.py:344]: **question:** What is the motivation for this change? It looks reasonable, I'mjust not sure I understand it.

File: lib/charms/rolling_ops/v0/rollingops.py [hunk 2]
```diff
@@ -393,6 +397,16 @@ def _on_acquire_lock(self: CharmBase, event: ActionEvent):
...
+
+        if self.name not in os.environ.get("JUJU_DISPATCH_PATH", ""):
+            # We can only guarantee to have the correct databag information when
+            # we are running in the {self.name}-relation. Given we can defer this
+            # event, we need to be careful we're only executing the run-with-lock
+            # when we have been actually granted the lock.
+            logger.warning(f"Tried to execute {event} outside of {self.name} context, defer")
+            event.defer()
+            return
+
```
Comment (marcoppenheimer) [inline lib/charms/rolling_ops/v0/rollingops.py:409]: **thought:** I was thinking here. The only reason we pass `event` to `callback` **IS** to defer it.
Why don't we just remove it? Don't pass event.

File: lib/charms/rolling_ops/v0/rollingops.py (hunk 1 above)
Comment (phvalguima) [reply lib/charms/rolling_ops/v0/rollingops.py:344]: I updated the main explanation of this PR. The thing we absolutely need to avoid is *reading* from the databag when we are outside of the `{self.name-relation-*` hooks. This method depends a lot on the databag state, so we need to enforce it.

Removing the whole event was just to clean up the code a bit and make this method tightly coupled with our `_peer_relation_changed`, which has the check.

File: lib/charms/rolling_ops/v0/rollingops.py (hunk 2 above)
Comment (phvalguima) [reply lib/charms/rolling_ops/v0/rollingops.py:409]: Likewise, we have a `Lock(self)`, which is a READ in the databag. We need this check then.

I thought about also removing this as an event. However, I want to keep users with the option of deferring this RunWithLock event. And I prefer them to defer RunWithLock (which is specific), then deferring RelationChanged, which is broad and does more tasks.
